src/python/pants/backend/python/dependency_inference/import_parser.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import ast as ast3
+import re
+import sys
+import warnings
+from dataclasses import dataclass
+from typing import Set
+
+from typed_ast import ast27
+
+from pants.util.ordered_set import FrozenOrderedSet
+from pants.util.strutil import ensure_text
+
+
+class ImportParseError(ValueError):
+    pass
+
+
+@dataclass(frozen=True)
+class ParsedPythonImports:
+    """All the discovered imports from a Python source file.
+
+    Explicit imports are imports from `import x` and `from module import x` statements. Inferred
+    imports come from strings that look like module names, such as
+    `importlib.import_module("example.subdir.Foo")`.
+    """
+
+    explicit_imports: FrozenOrderedSet[str]
+    inferred_imports: FrozenOrderedSet[str]
+
+
+def parse_file(source_code: str, *, module_name: str):
+    try:
+        # NB: The Python 3 ast is generally backwards-compatible with earlier versions. The only
+        # breaking change is `async` `await` becoming reserved keywords in Python 3.7 (deprecated
+        # in 3.6). If the std-lib fails to parse, we could use typed-ast to try parsing with a
+        # target version of Python 3.5, but we don't because Python 3.5 is almost EOL and has very
+        # low usage.
+        # We will also fail to parse Python 3.8 syntax if Pants is run with Python 3.6 or 3.7.
+        # There is no known workaround for this, beyond users changing their `./pants` script to
+        # always use >= 3.8.
+        tree = ast3.parse(source_code)
+        visitor_cls = _Py3AstVisitor if sys.version_info[:2] < (3, 8) else _Py38AstVisitor
+        return tree, visitor_cls
+    except Exception as e:
+        try:
+            return ast27.parse(source_code), _Py27AstVisitor
+        except Exception:
+            raise ImportParseError(f"Failed to parse source code for {module_name}:\n{e!r}")
```

Comment:
If this message ends up being displayed to the user, then it is a very bad UX.
When there is a syntax error, more information needs to be given, just saying that there is a syntax error somewhere in my_big_python_file.py is bad UX.

Reply:
<img width="750" alt="Screen Shot 2020-05-30 at 6 57 12 AM" src="https://user-images.githubusercontent.com/1268088/83330276-cc754480-a242-11ea-802c-1fb8a5881974.png">
vs:
<img width="1393" alt="Screen Shot 2020-05-30 at 6 57 42 AM" src="https://user-images.githubusercontent.com/1268088/83330287-de56e780-a242-11ea-84c4-3782138dd2a0.png">

Reply:
this might be helpful to adapt to pants.
https://github.com/pytest-dev/pytest/blob/0a03217903c21570d0b32ab052509fafbcc88ffc/src/_pytest/python.py#L517

Reply:
also, there should be a couple of tests that cover failure cases.

Reply:
We might also consider completely ignoring parse failures in the context of inference, in order to let the tool that folks were _trying_ to run tell them that the code doesn't parse.